File: apps/api/app/api/v1/routes/gallery.py
from __future__ import annotations

from dataclasses import asdict
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, HTTPException, Response, UploadFile, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.admin import require_admin
from app.db.session import get_session
from app.models.admin_user import AdminUser
from app.models.gallery_image import GalleryImage
from app.schemas.gallery_image import GalleryImageRead
from app.services.image_uploads import (
    create_uploaded_image,
    delete_uploaded_image_files,
    rotate_uploaded_image,
)

router = APIRouter()


def _delete_gallery_image_files(image: GalleryImage) -> None:
    delete_uploaded_image_files(
        storage_path=image.storage_path,
        thumbnail_storage_path=image.thumbnail_storage_path,
        tiny_thumbnail_storage_path=image.tiny_thumbnail_storage_path,
    )


@router.get("", response_model=list[GalleryImageRead])
async def list_gallery_images(
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[GalleryImageRead]:
    stmt = select(GalleryImage).order_by(GalleryImage.created_at.desc(), GalleryImage.id.desc())
    images = (await session.scalars(stmt)).all()
    return [GalleryImageRead.model_validate(image) for image in images]


@router.post("", response_model=list[GalleryImageRead], status_code=status.HTTP_201_CREATED)
async def upload_gallery_images(
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
    files: Annotated[list[UploadFile], File(...)],
    resize_mode: Annotated[str, Form()] = "keep",
    resize_width: Annotated[int | None, Form()] = None,
    resize_height: Annotated[int | None, Form()] = None,
) -> list[GalleryImageRead]:
    if not files:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="No files were uploaded"
        )

    if resize_mode not in {"keep", "resize"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resize mode")

    if resize_mode == "resize" and (
        resize_width is None or resize_height is None or resize_width <= 0 or resize_height <= 0
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Resize width and height must be positive integers",
        )

    created_images: list[GalleryImage] = []
    try:
        for file in files:
            uploaded = await create_uploaded_image(
                upload=file,
                resize_mode=resize_mode,
                resize_width=resize_width,
                resize_height=resize_height,
                storage_prefix="gallery",
            )
            image = GalleryImage(**asdict(uploaded))
            session.add(image)
            created_images.append(image)

        await session.commit()
    except Exception:
        await session.rollback()
        for image in created_images:
            _delete_gallery_image_files(image)
        raise

    for image in created_images:
        await session.refresh(image)

    return [GalleryImageRead.model_validate(image) for image in created_images]


@router.delete("/{image_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_gallery_image(
    image_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> Response:
    image = await session.get(GalleryImage, image_id)
    if image is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Gallery image not found")

    _delete_gallery_image_files(image)
    await session.delete(image)
    await session.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.patch("/{image_id}/rotate", response_model=GalleryImageRead)
async def rotate_gallery_image(
    image_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> GalleryImageRead:
    image = await session.get(GalleryImage, image_id)
    if image is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Gallery image not found")

    rotated = rotate_uploaded_image(
        storage_path=image.storage_path,
        thumbnail_storage_path=image.thumbnail_storage_path,
        tiny_thumbnail_storage_path=image.tiny_thumbnail_storage_path,
        image_url=image.image_url,
        thumbnail_url=image.thumbnail_url,
        tiny_thumbnail_url=image.tiny_thumbnail_url,
        content_type=image.content_type,
        original_filename=image.original_filename,
        gps_latitude=image.gps_latitude,
        gps_longitude=image.gps_longitude,
    )

    for field, value in asdict(rotated).items():
        setattr(image, field, value)

    await session.commit()
    await session.refresh(image)
    return GalleryImageRead.model_validate(image)

File: apps/web/src/lib/gallery.ts
export type GalleryImageRead = {
  id: number;
  image_url: string;
  thumbnail_url: string;
  tiny_thumbnail_url: string | null;
  width: number;
  height: number;
  thumbnail_width: number;
  thumbnail_height: number;
  tiny_thumbnail_width: number | null;
  tiny_thumbnail_height: number | null;
  content_type: string;
  original_filename: string | null;
  gps_latitude: number | null;
  gps_longitude: number | null;
  created_at: string;
};

function getApiBaseUrl() {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    throw new Error("Missing NEXT_PUBLIC_API_BASE_URL");
  }
  return baseUrl;
}

async function request<T>(path: string, token: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${getApiBaseUrl()}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      ...init?.headers,
    },
    cache: "no-store",
  });

  if (res.status === 401 || res.status === 403) {
    throw new Error("AUTH_REQUIRED");
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `Request failed: ${res.status}`);
  }

  if (res.status === 204) {
    return undefined as T;
  }

  return (await res.json()) as T;
}

async function requestMultipart<T>(
  path: string,
  token: string,
  body: FormData,
  init?: RequestInit,
): Promise<T> {
  const res = await fetch(`${getApiBaseUrl()}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      ...init?.headers,
    },
    body,
    cache: "no-store",
  });

  if (res.status === 401 || res.status === 403) {
    throw new Error("AUTH_REQUIRED");
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `Request failed: ${res.status}`);
  }

  return (await res.json()) as T;
}

export function listGalleryImages(token: string) {
  return request<GalleryImageRead[]>("/api/v1/gallery", token);
}

export function uploadGalleryImages(
  token: string,
  payload: {
    files: File[];
    resizeMode: "keep" | "resize";
    resizeWidth: number | null;
    resizeHeight: number | null;
  },
) {
  const body = new FormData();
  for (const file of payload.files) {
    body.append("files", file);
  }
  body.append("resize_mode", payload.resizeMode);
  if (payload.resizeMode === "resize") {
    body.append("resize_width", String(payload.resizeWidth ?? ""));
    body.append("resize_height", String(payload.resizeHeight ?? ""));
  }

  return requestMultipart<GalleryImageRead[]>("/api/v1/gallery", token, body, {
    method: "POST",
  });
}

export function deleteGalleryImage(token: string, imageId: number) {
  return request<void>(`/api/v1/gallery/${imageId}`, token, {
    method: "DELETE",
  });
}

export function rotateGalleryImage(token: string, imageId: number) {
  return request<GalleryImageRead>(`/api/v1/gallery/${imageId}/rotate`, token, {
    method: "PATCH",
  });
}
